README.md

```diff
@@ -20,7 +20,99 @@ The hope is to make it convenient to share visualizations between tools.
+Rendering visualizations should be possible in both Clojure and ClojureScript.
+Viewers may need to be tailored to different usage scenarios.
+
+## What are notes?
+
+The purpose of this library is to take a `note` or `notebook` consisting of `notes`,
+and produce markdown, hiccup, or html for displaying them.
+A note is a map containing either a value, form, or both.
+
+Example note:
+
+```clojure
+{:form (+ 1 2)
+ :value 3}
+```
+
+Example notebook:
+
+```clojure
+{:notes [{:value 3} {:value "Hello world"}]}
+```
+
+A note may be annotated for visualization with kindly:
+
+```clojure
+{:value (kind/hiccup [:svg [:circle {:r 10}]])}
+```
+
+## Usage
+
+No releases are available yet, please use a git dependency if you want to try it.
+
+```clojure
+(require '[scicloj.kindly-render.note.to-hiccup :as to-hiccup])
+(to-hiccup/render my-note)
+```
+
+Returns a hiccup representation of the visualization of `my-note`.
+
+```clojure
+(require '[scicloj.kindly-render.notes.to-html-page :as to-html-page])
+(to-html-page/notes-to-html my-notebook)
+```
+
+Returns a string of HTML representation of the `my-notebook`.
+
+See [examples/basic/page.clj](examples/basic/page.clj) for a concrete example.
+
+## Design
+
+Sometimes we want to produce a JavaScript enabled visualization, but sometimes we can't. Maybe we are making a PDF, and so we need to use a static image instead.
+
+Under `scicloj.kindly-render.note` are several targets:
+
+* `to-hiccup` (plain)
+* `to-hiccup-js` (javascript enhanced)
+* `to-markdown`
+* `to-scittle-reagent`
+
+The main entry point for each is `render`.
+
+So when making a PDF from HTML, users would call `to-hiccup/render` and get an image, but when making a rich web page they might call `to-hiccup-js/render` and get the JavaScript enhanced HTML instead.
+
+`render` first uses `kindly-advice` to ensure that the input `note` has a completed value, and has been checked for `kindly` annotations as metadata. At this point the `note` will contain a key `:kind` which indicates the visualization requested (if any). The completed note is then passed to `render-advice` which is a multimethod that dispatches on the `:kind` to produce the target output.
+
+### A multi-method per target
+
+Each target has a multi-method called `render-advice` with methods defined for as many `:kinds` as are supported by that target.
+
+Using multi-methods allows users to use the standard features of Clojure to add or replace viewers if they would like to.
+
+#### Fallback
+
+Each multi-method has a `:default` implementation.
+If no viewer is matched in `markdown`, it will fall back to `hiccup-js`. If no viewer is matched there, then it will fall back to `hiccup`.
+Fallback only happens in one direction because `hiccup` cannot fall back to `markdown`, but `markdown` can fall back to `hiccup`. Similarly plain `hiccup` cannot fall back to `hiccup-js`, but hiccup-js can fall back to `hiccup`.
```

Comment:
`hiccup` can fall back to `markdown`, since we can convert markdown to hiccup.

This is an important detail of the Clay implementation. I don't think we want to lose it.

Reply:
Rendering a visualization (note) to HTML cannot return Markdown text (that would make the HTML document invalid).
However if the visualization (note) itself represents Markdown `(kind/md "#hello")` then that is fine, it will be rendered from Markdown to HTML by the hiccup renderer.
I believe Clay must work like this -- if not, could you provide an example?

Reply:
In Clay, `hiccup` falls back to `markdown` by converting the markdown to hiccup. This is currently done using the `nextjournal/markdown` library, which is used in `scicloj.clay.v2.prepare/md->hiccup`.

One useful example is a `tech.ml.dataset` dataset. It is printed as Markdown table, which is converted to hiccup if we need to render an html target.

Reply:
Yes I see what you mean.
I want to make the md->hiccup explicit in those cases.
Here is a change to make dataset work the way you describe:
https://github.com/scicloj/kindly-render/pull/6
Instead of calling the top level markdown render, that kind is known to need conversion so we just do it right there.
This approach requires these cases to be explicitly handled, which is a bit of code duplication (we could extract the common part if that becomes a problem) - but I think it makes the responsibilities clearer.
We can discuss further if this is the right approach.

Reply:
I think there may be other cases we need to handle, but reviewing the known kinds I didn't spot any, so please let me know if there are others that need this.